shared/metadata.go

```diff
@@ -20,3 +50,94 @@ type MetadataLink struct {
 	TestPath string `yaml:"test"`
 	URL      string
 }
+
+func (m MetadataResults) Len() int           { return len(m) }
+func (m MetadataResults) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+func (m MetadataResults) Less(i, j int) bool { return m[i].String() < m[j].String() }
+
+func (m MetadataResult) String() string {
+	var urlString string
+	for _, url := range m.URLs {
+		urlString += url
+	}
+	return m.Test + urlString
+}
+
+// GetMetadataResponse retrieves the response to a WPT Metadata query.
+func GetMetadataResponse(testRuns []TestRun, client *http.Client) MetadataResponse {
+	metadataByteMap, err := util.CollectMetadata(client)
+	if err != nil {
+		return MetadataResponse{}
+	}
+	metadata := parseMetadata(metadataByteMap)
+	return constructMetadataResponse(testRuns, metadata)
+}
+
+// parseMetadata collects and parses all META.yml files from
+// wpt-metadata reposiroty.
+func parseMetadata(metadataByteMap map[string][]byte) map[string]Metadata {
+	var metadataMap = make(map[string]Metadata)
+	for path, data := range metadataByteMap {
+		var metadata Metadata
+		err := yaml.Unmarshal(data, &metadata)
+		if err != nil {
+			log.Warningf("Failed to unmarshal %s.yml.", path)
+			continue
+		}
+		metadataMap[path] = metadata
+	}
+	return metadataMap
+}
+
+// ConstructMetadataResponse constructs the response to a WPT Metadata query.
+// When parsing 'link' nodes, assume there is no mising information nor duplicates;
+// assume each test for each browser type is only associated with one bug.
+func constructMetadataResponse(testRuns []TestRun, metadata map[string]Metadata) MetadataResponse {
+	res := MetadataResults{}
+	for folderPath, data := range metadata {
+		testMap := make(map[string][]string)
+
+		for _, link := range data.Links {
+			var urls []string
+
+			var fullTestName = ""
+			if strings.HasPrefix(link.TestPath, folderPath) {
+				fullTestName = link.TestPath
+			} else {
+				fullTestName = path.Join(folderPath, link.TestPath)
+			}
+
+			if _, ok := testMap[fullTestName]; !ok {
+				testMap[fullTestName] = make([]string, len(testRuns))
+			}
+			urls = testMap[fullTestName]
+
+			for i, run := range testRuns {
+				// Only matches browser type.
```

Comment:
Why?

Reply:
If I understand what we discussed in the meeting correctly, there is no versioning in the Metadata YML files - Browser version or WPT Test Version (everything against the tip of the tree). When users query metadata against browsers, we will simply return the bug that fulfills the browser type. I guess there are two implications

1) I would assume its user's responsibility to check the browser version this bug files against;
2) User won't file the same bug against two different versions of a same browser.

Well obviously it is a big flaw, I think I do need to flash out more details on this issue.

Reply:
Instead of matching only the browser part of the product spec, I'd suggest encouraging users not to put versions in the product specs when they are annotating. Essentially, we leave the option up to the users. If they have legitimate reasons to include browser versions in META.yml, we should let them do so.

Reply:
> Also, sorry I might have asked this before, but do we have a doc for the format of `META.yml` files? For example, whether the `test` field is a relative path?

Right its right here (search for metadata format header) https://docs.google.com/document/d/1oWYVkc2ztANCGUxwNVTQHlWV32zq6Ifq9jkkbYNbSAg/edit#heading=h.7nki9mck5t64

Reply:
I modified the format of META.yml slightly and paste to the https://github.com/web-platform-tests/wpt-metadata/blob/master/README.

- I set test name field as mandatory 
- When a browser name is not specified, it will match to all browsers